src/librustc/middle/resolve_lifetime.rs

```diff
@@ -1475,12 +1475,61 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                         _ => None,
                     } {
                         debug!("id ={:?} span = {:?} name = {:?}", node_id, span, name);
-                        self.tcx.struct_span_lint_node(
+                        let mut err = self.tcx.struct_span_lint_node(
                             lint::builtin::UNUSED_LIFETIMES,
                             id,
                             span,
                             &format!("lifetime parameter `{}` never used", name)
-                        ).emit();
+                        );
+                        if let Some(parent_def_id) = self.tcx.parent(def_id) {
+                            if let Some(node_id) = self.tcx.hir.as_local_node_id(parent_def_id) {
+                                if let Some(Node::Item(hir_item)) = self.tcx.hir.find(node_id) {
+                                    match hir_item.node {
+                                        hir::ItemKind::Fn(_, _, ref generics, _) |
+                                        hir::ItemKind::Impl(_, _, _, ref generics, _, _, _) => {
+                                            let unused_lt_span = if generics.params.len() == 1 {
```

Comment:
and maybe the code to get the `unused_lt_span` would be a nice helper, too? it doesn't seem to use too many free variables